### ISSUE TITLE
Issue with System.IO.GetIsCaseSensitive

### ISSUE THREAD

**Issue by BruceForstall** [dotnet/runtime#54313]
I was running a crossgen2 of System.Private.CoreLib.dll (actually a SuperPMI collection of that), and getting a failure. I put it under the VS debugger and got something completely different, namely:

```
Process terminated. Assertion failed.
Casing test failed: System.UnauthorizedAccessException: Access to the path 'C:\WINDOWS\CASESENSITIVETEST3e2cf88a10694adbaeb85ab4ec42540f' is denied.
   at System.IO.Strategies.FileStreamHelpers.ValidateFileHandle(SafeFileHandle fileHandle, String path, Boolean useAsyncIO) in System.Private.CoreLib.dll:token 0x6005d7a+0x27
   at System.IO.Strategies.FileStreamHelpers.CreateFileOpenHandle(String path, FileMode mode, FileAccess access, FileShare share, FileOptions options, Int64 preallocationSize) in System.Private.CoreLib.dll:token 0x6005d74+0x83
   at System.IO.Strategies.WindowsFileStreamStrategy..ctor(String path, FileMode mode, FileAccess access, FileShare share, FileOptions options, Int64 preallocationSize) in System.Private.CoreLib.dll:token 0x6005e36+0x23
   at System.IO.Strategies.FileStreamHelpers.ChooseStrategyCore(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options, Int64 preallocationSize) in System.Private.CoreLib.dll:token 0x6005d71+0x28
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options, Int64 preallocationSize) in System.Private.CoreLib.dll:token 0x60059dc+0x10e
   at System.IO.PathInternal.GetIsCaseSensitive() in System.Private.CoreLib.dll:token 0x6005ace+0x27
   at System.IO.PathInternal.GetIsCaseSensitive() in System.Private.CoreLib.dll:token 0x6005ace+0x27
   at System.IO.PathInternal..cctor() in System.Private.CoreLib.dll:token 0x6005adb+0x0
   at System.IO.Enumeration.FileSystemEnumerableFactory.MatchesPattern(String expression, ReadOnlySpan`1 name, EnumerationOptions options) in System.Private.CoreLib.dll:token 0x6005e92+0x0
   at System.IO.Enumeration.FileSystemEnumerableFactory.<>c__DisplayClass3_0.<UserFiles>b__1(FileSystemEntry& entry) in System.Private.CoreLib.dll:token 0x6005e9b+0x8
   at System.IO.Enumeration.FileSystemEnumerator`1.MoveNext() in System.Private.CoreLib.dll:token 0x6005e79+0x0
   at Internal.CommandLine.Helpers.AppendExpandedPaths(Dictionary`2 dictionary, String pattern, Boolean strict) in crossgen2.dll:token 0x600011a+0x8c
   at ILCompiler.Program.ProcessCommandLine(String[] args) in crossgen2.dll:token 0x6000156+0x129
   at ILCompiler.Program.Run(String[] args) in crossgen2.dll:token 0x6000159+0x6
   at ILCompiler.Program.Main(String[] args) in crossgen2.dll:token 0x6000161+0x5
```

This comes from https://github.com/dotnet/runtime/blob/01b7e73cd378145264a7cb7a09365b41ed42b240/src/libraries/Common/src/System/IO/PathInternal.CaseSensitivity.cs#L55

The issue is that my debugging environment didn't include TEMP (you can see above it's using "C:\WINDOWS" as the "temp" directory). When I added "TEMP=C:\Users\brucefo\AppData\Local\Temp" to my environment, it worked fine.

Question: Should this "Debug.Fail" exist? The comment says:
```
                // In case something goes terribly wrong, we don't want to fail just because
                // of a casing test, so we assume case-insensitive-but-preserving.
```
so should it just "move on" without this fatal error?

**Comment by BruceForstall** [comment]
@stephentoub

**Comment by MichalStrehovsky** [comment]
One can ask Windows to be case sensitive per directory (`fsutil.exe file setCaseSensitiveInfo`).

Do we really treat case sensitivity of all paths based on whether %TEMP% is case sensitive?

**Comment by stephentoub** [comment]
> Do we really treat case sensitivity of all paths based on whether %TEMP% is case sensitive?

For decades since the beginning of .NET it's assumed that a Windows file system was case-insensitive, and has never checked per directory.  This GetIsCaseSensitive dynamic check was added primarily in support of cross-platform.  While there have been a few requests over the years for something more granular, e.g. https://github.com/dotnet/runtime/issues/34235, it's never risen to the level where we want to pay for a syscall per directory in file path manipulations, comparisons, etc. to actually do anything about it.  On Windows, I can count on one hand I think the number of folks I've heard about actually having this cause a problem.

What is it you're proposing be done?

> Should this "Debug.Fail" exist?

This is the first time I'm aware of someone hitting it (and I'm amazed you're not hitting other issues with your temporary path pointing to c:\Windows).  But, sure, we can delete the assert, given you were actually able to trigger it.

**Comment by MichalStrehovsky** [comment]
> What is it you're proposing be done?

To me, assume case-preserving and case sensitive on Windows without looking at %TEMP% sounds the easiest to reason about.

**Comment by stephentoub** [comment]
> To me, assume case-preserving and case insensitive on Windows without looking at %TEMP% sounds the easiest to reason about.

Ah, you're not asking for it to be more dynamic, you're asking for it to be less :-)

**Comment by MichalStrehovsky** [comment]
> Ah, you're not asking for it to be more dynamic, you're asking for it to be less :-)

Yeah, checking by creating a file in TEMP feels rather arbitrary and pretty "dirty". It feels like we would be better off not doing it.

**Comment by lambdageek** [comment]
> This GetIsCaseSensitive dynamic check was added primarily in support of cross-platform. While there have been a few requests over the years for something more granular, e.g. #34235, it's never risen to the level where we want to pay for a syscall per directory in file path manipulations, comparisons, etc. to actually do anything about it. 

On Android, `GetIsCaseSensitive` is both a perf bottleneck (see https://github.com/dotnet/maui/issues/822#issuecomment-863323985) and wrong: while the internal fs on an Android device is usually ext4, which is usually case sensitive - sd cards are usually FAT and usually case-insensitive.  So checking the temp folder doesn't actually give a useful result for all cases.  At the same time, doing I/O from a static constructor means we slow down app startup (a crucial metric on low-powered Android devices).

/cc @grendello

**Comment by grendello** [comment]
This case-sensitivity check, if it's necessary, should be done very lazily - whenver the application itself accesses a given filesystem for the first time. However, I would say that on Android devices it would be very safe to assume case-sensitivity, even with exFAT, that we're dealing with a case-sensitive filesystem. An option should be provided to perform the check, off by default IMO.

**Comment by grendello** [comment]
It should be possible to look at `/proc/mounts` to find out what options are for the fs mounted on or above the path being accessed, if we need to have such a check in place. The `TEMP` check is not only wasteful but also can yield false positives/negatives - since any directory on Unix can be a mount point of another operating system, so such checks would have to be performed on every first access to any mounted filesystem, kind of costly (and unnecessary).

**Comment by stephentoub** [comment]
We can go back to hardcoding a value per OS if that's more desirable.

We've always known it's not 100% accurate, whether hardcoded or dynamically generated once per process.  But doing the 100% correct thing, with casing potentially different on a directory by directory basis, is both super expensive and very error prone, so it was a conscious tradeoff.

**Comment by grendello** [comment]
I think a hardcoded value is a perfectly valid bet here.

**Comment by lambdageek** [comment]
Made an issue https://github.com/dotnet/runtime/issues/54339 and PR https://github.com/dotnet/runtime/pull/54340 for Android.

> doing the 100% correct thing, with casing potentially different on a directory by directory basis, is both super expensive and very error prone,

I wonder if we can give application writers (1) an API to do the work themselves (like https://github.com/dotnet/runtime/issues/34235 suggested), (2) an AppContext switch to select from some reasonable set of behaviors for the BCL (eg: use a constant guess based on platform / use a runtime checked value / check per-path in an OS-specific way)

**Comment by stephentoub** [comment]
> I wonder if we can give application writers (1) an API to do the work themselves (like #34235 suggested), (2) an AppContext switch to select from some reasonable set of behaviors for the BCL (eg: use a constant guess based on platform / use a runtime checked value / check per-path in an OS-specific way)

An AppContext switch to override the default hardcoded value might be reasonable _if_ we have evidence it'll actually be used.  But the last piece ("check per-path in an OS-specific way") is a non-trivial dev effort and would really need to be justified.

**Comment by grendello** [comment]
@stephentoub on iOS you can use the hard-coded default blindly, it's always gonna be case-aware and case-insensitive. On Android, case-aware and case-sensitive is currently a safe bet too, but if need be, settings can be gleamed from`/proc/mounts` and fs mount options - it should be then a simple check of the parent ifs of a given path, that information can be cached (mount points) since mounts won't change during application's run time. Currently, however, it's an overkill.

**Comment by adamsitnik** [comment]
The issue described here by @BruceForstall got solved by @lambdageek in #54340

Once our customers ask for it, we should consider making it configurable. For now, I am just going to close it.
